### PR TITLE
Ensure node-based debug adapters spawn same node exec as Theia

### PR DESCRIPTION
**Pull request by perspectivus1**
Signed-off-by: Eyal Barlev <eyal.barlev@sap.com>

Sometimes the container/machine running Theia has more than one node executable available.

The node executable used to run Theia should also be used to run node-based debug adapters; whereas, another node executable can be used to run/debug user applications.

Debug adapters specify their runtime in their corresponding `package.json` file (e.g. [link](https://code.visualstudio.com/api/extension-guides/debugger-extension#anatomy-of-the-package.json-of-a-debugger-extension)):
```json
"contributes": {
  "debuggers": [
    {
       "runtime": "node"
```

The debug adapter typically does not specify the full path of the node executable, which means that Theia uses the first node executable on the `PATH` environment variable to spawn the debug adapter. However, this might not be the same node executable that was used to run Theia.

For us at SAP, this is critical, as we expose node executables to the Theia container, and those executables might run remotely on other containers that have no access to the debug adapter code (which resides on the Theia container).

This pull request ensures that if a debug adapter doesn't specify a fully qualified node runtime, Theia would spawn the debug adapter using the same node executable that was used to launch Theia.

This change should not adversely affect existing flows, but rather add support for the requirement detailed above.

### CODE REVIEW

**Comment by akosyakov** [comment]
We've discussed something like that for TS language servers already. cc @marcdumais-work @kittaakos 

I believe the proper there and here is to use `child_process.fork` instead of guessing `node` executable from args. It's designed to start new processes with exactly the same node which is used to run Theia.

Why guessing would be better? @tolusha What is your take on it?

**Comment by akosyakov** [comment]
See VS Code use `fork` as well: https://github.com/microsoft/vscode/blob/466c132f5b4362b3be485a1f7a624b3a7316ddb8/src/vs/workbench/contrib/debug/node/debugAdapter.ts#L186-L205

**Comment by perspectivus1** [comment]
I tried your proposed approach [here](https://github.com/perspectivus1/theia/blob/fork2/packages/debug/src/node/vscode/vscode-debug-adapter-contribution.ts#L234-L240). This indeed ran the debug adapter using `child_process.fork()`, but debugging user code kept failing with this error message:
```
Starting inspector on 127.0.0.1:24166 failed: address already in use
```

Do you know why this is happening?

Note that the node's API [documentation](https://nodejs.org/api/child_process.html#child_process_child_process_fork_modulepath_args_options) specifically states this:

> `child_process.fork()` will spawn new Node.js instances using the **`process.execPath`** of the parent process.

**Comment by tolusha** [comment]
I personally stand for `fork`

**Comment by akosyakov** [comment]
@perspectivus1 Can you open an alternative PR with `fork` that we can have a look as well?

**Comment by perspectivus1** [comment]
I created an alternative PR (#5508), with a `fork()` implementation (that fails to debug user code).

**Comment by kittaakos** [comment]
This won't work when running from electron, as it will spawn/fork a new electron process. Also, we have to generalize it and reuse for the TS and JSON LSs.

**Comment by akosyakov** [comment]
We should sort out Electron before merging.

**Comment by perspectivus1** [comment]
@akosyakov, I fixed the Electron case.

@kittaakos, there's a simple change that I have not pushed, that when running Electron would **spawn the debug adapter as a node process**:

In file `packages\debug\src\node\debug-adapter-factory.ts`:
```javascript
        const options = { stdio: ['pipe', 'pipe', 2], env: environment.electron.runAsNodeEnv() };
```

Should I add it to this pull request?

File: packages/debug/src/node/debug-adapter-factory.ts [hunk 1]
```diff
@@ -66,6 +68,12 @@ export class LaunchBasedDebugAdapterFactory implements DebugAdapterFactory {
         const isForkOptions = (forkOptions: RawForkOptions | any): forkOptions is RawForkOptions =>
             !!forkOptions && !!forkOptions.modulePath;
 
+        if (!isForkOptions(executable)) {
+            if ((executable as DebugAdapterSpawnExecutable).command === 'node' && !environment.electron.is()) {
```
Comment (kittaakos) [inline packages/debug/src/node/debug-adapter-factory.ts]: From the [description](https://github.com/theia-ide/theia/pull/5492#issue-288783989):
> The node executable used to run Theia should also be used to run node-based debug adapters

This won't be the case if you run in electron.

File: packages/debug/src/node/debug-adapter-factory.ts [hunk 2]
```diff
@@ -66,8 +68,18 @@ export class LaunchBasedDebugAdapterFactory implements DebugAdapterFactory {
         const isForkOptions = (forkOptions: RawForkOptions | any): forkOptions is RawForkOptions =>
             !!forkOptions && !!forkOptions.modulePath;
 
+        const options: { stdio: (string|number)[], env?: object } = { stdio: ['pipe', 'pipe', 2] };
+        if (!isForkOptions(executable)) {
+            if ((executable as DebugAdapterSpawnExecutable).command === 'node') {
+                if (environment.electron.is()) {
+                    options.env = environment.electron.runAsNodeEnv();
```
Comment (perspectivus1) [inline packages/debug/src/node/debug-adapter-factory.ts]: @kittaakos, is this what you meant?

Comment (kittaakos) [reply packages/debug/src/node/debug-adapter-factory.ts]: It looks much better now, logically it is correct.

Can we move this Node.js-specific customization to the process-factory?
What's your take on that, @akosyakov?

Comment (akosyakov) [reply packages/debug/src/node/debug-adapter-factory.ts]: I wonder why `execPath` is used only in `else` branch.

Comment (akosyakov) [reply packages/debug/src/node/debug-adapter-factory.ts]: > Can we move this Node.js-specific customization to the process-factory?
What's your take on that, @akosyakov?

Do you mean if `command` ===  `process.execPath`? for `node` i don't think so, it is not the same

I don't think this code should use process factory at all, but it is another issue. User processes (running in user dev env like terminals and task) should be created by a factory, all other should be created vis node js APIs.

Comment (kittaakos) [reply packages/debug/src/node/debug-adapter-factory.ts]: True.

Comment (perspectivus1) [reply packages/debug/src/node/debug-adapter-factory.ts]: > I wonder why `execPath` is used only in `else` branch.

In the case of Electron, `execPath` contains the full path of the Electron executable; whereas we want to spawn Electron's embedded node executable. For that we set the environment variable `ELECTRON_RUN_AS_NODE`, provided by [`ElectronDev.runAsNodeEnv()`](https://github.com/perspectivus1/theia/blob/master/dev-packages/application-package/src/environment.ts#L64).

Comment (perspectivus1) [reply packages/debug/src/node/debug-adapter-factory.ts]: > Can we move this Node.js-specific customization to the process-factory?

So, should I move the implementation elsewhere?

Comment (kittaakos) [reply packages/debug/src/node/debug-adapter-factory.ts]: > `execPath` contains the full path of the Electron executable; whereas we want to spawn Electron's embedded node executable.

I am not sure if it is true. But please correct me if I am wrong.
If I have Node.js `v10.15.3` available from the `PATH` and I start a Theia-based electron application which comes with electron `3.1.7` (as of today), then I will run with a different Node.js version than I am expecting.
```
Akoss-MacBook-Pro:theia akos.kitta$ npx electron -i
> console.log(process.versions.node)
10.2.0
undefined
> .exit
```

Does this make sense?

Comment (kittaakos) [reply packages/debug/src/node/debug-adapter-factory.ts]: > we want to spawn Electron's embedded node executable.
This part is not true, I think.

Comment (perspectivus1) [reply packages/debug/src/node/debug-adapter-factory.ts]: When Theia runs in node, we can control the node executable that is used to spawn the debug adapter, because we can infer it from `process.execPath`.

When Theia runs in Electron, `process.execPath` contains a full path to the electron executable. Therefore, we have 2 choices:
1. **Spawn** a node process using the first node executable found in the `PATH` environment variable.
2. **Fork** an electron process using the `ELECTRON_RUN_AS_NODE` environment variable, which will use Electron's embedded node.

However, the current implementation of `AbstractVSCodeDebugAdapterContribution.provideDebugAdapterExecutable()` creates a `DebugAdapterExecutable` that causes Theia to **spawn**, rather than fork debug adapters.

In other words, in the case of a node-based debug adapter, the code above always spawn a **node** process, regardless of whether Theia runs in Electron or in node and regardless of the value of the `ELECTRON_RUN_AS_NODE` environment variable.

If this is the desired behavior, then this PR fulfills the requirement.

If we wish to use Electron's embedded node, I will have to modify PR #5508.

Please advise

Comment (akosyakov) [reply packages/debug/src/node/debug-adapter-factory.ts]: Forking is a proper solution. It is the same way as VS Code does it. I would be in favor of closing this PR and modifying #5508.

**Comment by akosyakov** [comment]
Closing it in favor of https://github.com/theia-ide/theia/pull/5508
